crates/matrix-sdk/src/room/joined.rs

```diff
@@ -644,64 +643,82 @@ impl Joined {
         &self,
         body: &str,
         content_type: &Mime,
-        mut reader: &mut R,
+        reader: &mut R,
         txn_id: Option<Uuid>,
     ) -> Result<send_message_event::Response> {
-        let (response, file) = if self.is_encrypted() {
-            #[cfg(feature = "encryption")]
+        #[cfg(feature = "encryption")]
+        async fn get_content_encrypted<R: Read>(
+            j: &Joined,
+            body: &str,
+            reader: &mut R,
+        ) -> Result<MessageType> {
             let mut reader = matrix_sdk_base::crypto::AttachmentEncryptor::new(reader);
-            #[cfg(feature = "encryption")]
             let content_type = &mime::APPLICATION_OCTET_STREAM;
 
-            let response = self.client.upload(content_type, &mut reader).await?;
+            let response = j.client.upload(content_type, &mut reader).await?;
 
-            #[cfg(feature = "encryption")]
-            let file: Option<Box<EncryptedFile>> = {
+            let file: EncryptedFile = {
                 let keys = reader.finish();
-                Some(Box::new(
-                    ruma::events::room::EncryptedFileInit {
-                        url: response.content_uri.clone(),
-                        key: keys.web_key,
-                        iv: keys.iv,
-                        hashes: keys.hashes,
-                        v: keys.version,
-                    }
-                    .into(),
-                ))
+                ruma::events::room::EncryptedFileInit {
+                    url: response.content_uri,
+                    key: keys.web_key,
+                    iv: keys.iv,
+                    hashes: keys.hashes,
+                    v: keys.version,
+                }
+                .into()
             };
-            #[cfg(not(feature = "encryption"))]
-            let file: Option<Box<EncryptedFile>> = None;
 
-            (response, file)
-        } else {
-            let response = self.client.upload(content_type, &mut reader).await?;
-            (response, None)
-        };
+            let res: Result<MessageType> = Ok(match content_type.type_() {
```

Comment:
This `let res` is redundant, let's just return directly like the other method does.

Reply:
Oops, I think that was left over when I tried it with closures instead.